README.md
<h1 align="center">CoCart Products <a href="https://github.com/co-cart/cocart-products/releases/latest/"><img src="https://img.shields.io/static/v1?goVersion=&message=v1.0.0&label=&color=9a6fc4&style=flat-square"></a></h1>

<p align="center"><img src="https://raw.githubusercontent.com/co-cart/co-cart/master/.github/Logo-1024x534.png.webp" alt="CoCart" /></p>

This premium add-on provides a public version of accessing products, categories, tags, attributes and even reviews from your WooCommerce store without the requirement of authentication.

Supports other products types created by WooCommerce extensions such as Subscriptions and more. Everything you need to display your products to your customers and add to the cart is all here.

<p align="center">
	<a href="https://wordpress.org/" target="_blank">
		<img src="https://img.shields.io/static/v1?label=&message=5.4+-+5.7&color=blue&style=flat-square&logo=wordpress&logoColor=white" alt="WordPress Versions">
	</a>
	<a href="https://woocommerce.com/" target="_blank">
		<img src="https://img.shields.io/static/v1?label=&message=4.3+-+5.4&color=96588A&style=flat-square&logo=woocommerce&logoColor=white" alt="WooCommerce Versions">
	</a>
	<a href="https://www.php.net/" target="_blank">
		<img src="https://img.shields.io/static/v1?label=&message=7.0+-+7.4&color=777bb4&style=flat-square&logo=php&logoColor=white" alt="PHP Versions">
	</a>
</p>

---
CoCart Products is a commercial plugin that is available with CoCart Pro from [https://cocart.xyz](https://cocart.xyz/pro/).

The plugin is hosted here on a public GitHub repository in order to better facilitate community contributions from developers and users alike.

If you have a suggestion, a bug report, or a patch for an issue, feel free to [submit it here](https://github.com/co-cart/cocart-products/issues/new/choose).

I do ask, however, that if you are using the plugin on a live site that you please purchase a valid license from the website. I cannot provide support to anyone that does not hold a valid license key.

Thank you.

---

## Looking for documentation?

Head over here: [https://docs.cocart.xyz/products.html](https://docs.cocart.xyz/products.html?utm_medium=github.com&utm_source=repository&utm_campaign=readme&utm_content=cocart)

## Features

 * Return all published products.
 * Return an individual product.
 * Return an individual variable product and all it's variations in one request.
 * Return all product categories.
 * Return all product tags.
 * Return all product attributes.
 * Return product attribute terms.
 * Return all product reviews.
 * Return an individual product review.
 * Create a product review. (Requires customer to use same email address used for ordering the product)
 * No authentication required.

Each API can be filtered to anything you require to display results for the front. There are no restrictions. If you need to add additional condidtions, there are filters that let you alter the query of the results before the query is made.

## Requirement

You will need to be using [CoCart v2.0](https://wordpress.org/plugins/cart-rest-api-for-woocommerce) or higher before installing this add-on.

## Installation

Download the latest release and upload to your WordPress dashboard via the plugins page and then activate.

## Contributing to CoCart Products

If you have a patch or have stumbled upon an issue with CoCart Products, you can contribute this back to the code. Please read the [contributor guidelines](https://github.com/co-cart/cocart-products/blob/master/.github/CONTRIBUTING.md) for more information how you can do this.

---

## License

[![License](https://img.shields.io/badge/license-GPL--3.0%2B-red.svg)](https://github.com/co-cart/cocart-products/blob/master/LICENSE.md)

CoCart Products is released under [GNU General Public License v3.0](http://www.gnu.org/licenses/gpl-3.0.html).

## Credits

CoCart Products is developed and maintained by [Sébastien Dumont](https://github.com/seb86).

---

[sebastiendumont.com](https://sebastiendumont.com) &nbsp;&middot;&nbsp;
GitHub [@seb86](https://github.com/seb86) &nbsp;&middot;&nbsp;
Twitter [@sebd86](https://twitter.com/sebd86)

<p align="center">
    <img src="https://raw.githubusercontent.com/seb86/my-open-source-readme-template/master/a-sebastien-dumont-production.png" width="353">
</p>

CHANGELOG.md
# Changelog for CoCart Products

## v1.0.0-beta.12 - 19th June, 2021

⚠️ Best to update **CoCart Products add-on** before updating CoCart to `v3.1`. This will help prevent crashing the site as the products API has now merged with CoCart. ⚠️

* Compatible: CoCart v3.1

## v1.0.0-beta.11 - 15th May, 2021

* NEW: Added orderby product name.
* NEW: Added orderby include posts in.
* NEW: Filter by catalog visibility and rating.
* Improved: Filter by many attributes and taxonomies instead of singular.
* Fixed: Permission issues with viewing taxonomies.
* Fixed: Multi-site support.

## v1.0.0-beta.10 - 10th August, 2020

* Change: Check if CoCart exists instead of CoCart Pro.
* Improved: Admin notices.
* Improved: Returned all gallery images in all image sizes.
* Improved: Added more product data to return and updated schema 

## v1.0.0-beta.9

* **NEW**: Added more orderby options.
* * Alphabetical
* * Reverse Alphabetical
* * By Stock
* * Review Count
* * On Sale First
* * Featured First
* * Price Ascending
* * Price Descending
* * Total Sales
* * Average Rating

* **NEW**: Added translation plugin action link on the plugins page.

## v1.0.0-beta.8

* **NEW**: Added subscription product data for subscription products
* Added `@licence` in each API file header.

## v1.0.0-beta.7

* Updated CoCart Pro not installed notice.

## v1.0.0-beta.6

* **NEW**: Added more orderby options.
* Fixed disable auto updates action.

## v1.0.0-beta.5

* Removed `tax_class` as a filterable option.
* Updated collection parameters.

## v1.0.0-beta.4

* **NEW**: Added full size and custom to available attachment sizes.
* Corrected error for invalid attribute.

## v1.0.0-beta.3

* Return featured image in all sizes.

## v1.0.0-beta.2

* Removed the need for context option.
* Renamed filter name `cocart_rest_prepare_{$this->taxonomy}` to `cocart_prepare_{$this->taxonomy}`.

## v1.0.0-beta.1 - 04th March, 2020

* Initial release.


includes/class-cocart-products.php
<?php
/**
 * CoCart Products core setup.
 *
 * @author   Sébastien Dumont
 * @category Package
 * @license  GPL-2.0+
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

final class CoCart_Products {

	/**
	 * Plugin Version
	 *
	 * @access public
	 * @static
	 */
	public static $version = '1.0.0-beta.12';

	/**
	 * Required WordPress Version
	 *
	 * @access public
	 * @static
	 */
	public static $required_wp = '5.4';

	/**
	 * Required WooCommerce Version
	 *
	 * @access public
	 * @static
	 */
	public static $required_woo = '4.3';

	/**
	 * Required PHP Version
	 *
	 * @access public
	 * @static
	 */
	public static $required_php = '7.0';

	/**
	 * Required CoCart Version
	 *
	 * @access public
	 * @static
	 */
	public static $required_cocart = '2.0.0';

	/**
	 * Initiate CoCart Products.
	 *
	 * @access public
	 * @static
	 */
	public static function init() {
		self::setup_constants();
		self::includes();

		// Install CoCart Products upon activation.
		register_activation_hook( COCART_PRODUCTS_FILE, array( __CLASS__, 'install_cocart_products' ) );

		// Update CoCart add-on counter upon deactivation.
		register_deactivation_hook( COCART_PRODUCTS_FILE, array( __CLASS__, 'deactivate_addon' ) );

		// Load translation files.
		add_action( 'init', array( __CLASS__, 'load_plugin_textdomain' ), 0 );

		// Load REST API.
		add_action( 'init', array( __CLASS__, 'load_rest_api' ) );
	} // END init()

	/**
	 * Return the name of the package.
	 *
	 * @access public
	 * @static
	 * @return string
	 */
	public static function get_name() {
		return 'CoCart Products';
	}

	/**
	 * Return the version of the package.
	 *
	 * @access public
	 * @static
	 * @return string
	 */
	public static function get_version() {
		return self::$version;
	}

	/**
	 * Return the path to the package.
	 *
	 * @access public
	 * @static
	 * @return string
	 */
	public static function get_path() {
		return dirname( __DIR__ );
	}

	/**
	 * Setup Constants
	 *
	 * @access public
	 * @static
	 */
	public static function setup_constants() {
		self::define( 'COCART_PRODUCTS_ABSPATH', dirname( COCART_PRODUCTS_FILE ) . '/' );
		self::define( 'COCART_PRODUCTS_PLUGIN_BASENAME', plugin_basename( COCART_PRODUCTS_FILE ) );
		self::define( 'COCART_PRODUCTS_VERSION', self::$version );
		self::define( 'COCART_PRODUCTS_SLUG', 'cocart-products' );
		self::define( 'COCART_PRODUCTS_URL_PATH', untrailingslashit( plugins_url( '/', COCART_PRODUCTS_FILE ) ) );
		self::define( 'COCART_PRODUCTS_FILE_PATH', untrailingslashit( plugin_dir_path( COCART_PRODUCTS_FILE ) ) );
		self::define( 'COCART_PRODUCTS_PLUGIN_URL', 'https://cocart.xyz/add-ons/products/' );
		self::define( 'COCART_PRODUCTS_REVIEW_URL', 'https://cocart.xyz/submit-review/?wpf15410_12=CoCart%20Products' );
		self::define( 'COCART_PRODUCTS_DOCUMENTATION_URL', 'https://docs.cocart.xyz/products.html' );
		self::define( 'COCART_PRODUCTS_TRANSLATION_URL', 'https://translate.cocart.xyz/projects/cocart-products/' );
	} // END setup_constants()

	/**
	 * Define constant if not already set.
	 *
	 * @access private
	 * @static
	 * @param  string      $name
	 * @param  string|bool $value
	 */
	private static function define( $name, $value ) {
		if ( ! defined( $name ) ) {
			define( $name, $value );
		}
	} // END define()

	/**
	 * Includes CoCart Products REST-API.
	 *
	 * @access public
	 * @static
	 * @return void
	 */
	public static function includes() {
		include_once COCART_PRODUCTS_ABSPATH . '/includes/class-cocart-products-autoloader.php';
		include_once COCART_PRODUCTS_ABSPATH . '/includes/class-cocart-products-helpers.php';
		require_once COCART_PRODUCTS_ABSPATH . '/includes/class-cocart-products-install.php';
	} // END includes()

	/**
	 * Install CoCart Products upon activation.
	 *
	 * @access public
	 * @static
	 */
	public static function install_cocart_products() {
		self::activation_check();

		CoCart_Products_Install::install();

		self::activate_addon();
	} // END install_cocart_products()

	/**
	 * Checks the server environment and other factors and deactivates the plugin if necessary.
	 *
	 * @access public
	 * @static
	 */
	public static function activation_check() {
		if ( ! CoCart_Products_Helpers::is_environment_compatible() ) {
			self::deactivate_plugin();
			wp_die( sprintf( __( '%1$s could not be activated. %2$s', 'cocart-products' ), 'CoCart Products', CoCart_Products_Helpers::get_environment_message() ) );
		}
	} // END activation_check()

	/**
	 * Deactivates the plugin if the environment is not ready.
	 *
	 * @access public
	 * @static
	 */
	public static function deactivate_plugin() {
		require_once ABSPATH . 'wp-admin/includes/plugin.php';

		deactivate_plugins( plugin_basename( COCART_PRODUCTS_FILE ) );

		if ( isset( $_GET['activate'] ) ) {
			unset( $_GET['activate'] );
		}
	} // END deactivate_plugin()

	/**
	 * Load Products.
	 *
	 * @access public
	 * @static
	 */
	public static function load_rest_api() {
		include_once COCART_PRODUCTS_ABSPATH . '/includes/class-cocart-products-init.php';
	} // END load_rest_api()

	/**
	 * Runs when the plugin is activated.
	 *
	 * Adds plugin to list of installed CoCart add-ons.
	 *
	 * @access public
	 */
	public static function activate_addon() {
		$addons_installed = get_option( 'cocart_addons_installed', array() );

		$plugin = plugin_basename( COCART_PRODUCTS_FILE );

		// Check if plugin is already added to list of installed add-ons.
		if ( ! in_array( $plugin, $addons_installed, true ) ) {
			array_push( $addons_installed, $plugin );
			update_option( 'cocart_addons_installed', $addons_installed );
		}
	} // END activate_addon()

	/**
	 * Runs when the plugin is deactivated.
	 *
	 * Removes plugin from list of installed CoCart add-ons.
	 *
	 * @access public
	 */
	public static function deactivate_addon() {
		$addons_installed = get_option( 'cocart_addons_installed', array() );

		$plugin = plugin_basename( COCART_PRODUCTS_FILE );

		// Remove plugin from list of installed add-ons.
		if ( in_array( $plugin, $addons_installed, true ) ) {
			$addons_installed = array_diff( $addons_installed, array( $plugin ) );
			update_option( 'cocart_addons_installed', $addons_installed );
		}
	} // END deactivate_addon()

	/**
	 * Load the plugin translations if any ready.
	 *
	 * Note: the first-loaded translation file overrides any following ones if the same translation is present.
	 *
	 * Locales found in:
	 *      - WP_LANG_DIR/cocart-products/cocart-products-LOCALE.mo
	 *      - WP_LANG_DIR/plugins/cocart-products-LOCALE.mo
	 *
	 * @access public
	 * @static
	 */
	public static function load_plugin_textdomain() {
		if ( function_exists( 'determine_locale' ) ) {
			$locale = determine_locale();
		} else {
			$locale = is_admin() ? get_user_locale() : get_locale();
		}

		$locale = apply_filters( 'plugin_locale', $locale, 'cocart-products' );

		unload_textdomain( 'cocart-products' );
		load_textdomain( 'cocart-products', WP_LANG_DIR . '/cocart-products/cocart-products-' . $locale . '.mo' );
		load_plugin_textdomain( 'cocart-products', false, plugin_basename( dirname( COCART_PRODUCTS_FILE ) ) . '/languages' );
	} // END load_plugin_textdomain()

} // END class
